# String manipulation

!!! note "Learning objectives"

    * Be comfortable with regular expressions in R
    * Use functions from package `stringr`

## Can't I change it in Excel? And other reasons to learn string manipulation.

Short answer: sure; long answer: *it depends*. If the data is small enough, Excel (or any spreadsheet software) is sufficient to fulfil your exploratory data analysis needs. However, for larger data sets, can you ensure that you have performed adequate, consistent, and robust data cleaning before analyses? Furthermore, if you are importing tabular data into R, you likely intend to perform statistical/numerical analyses and create a near-publication-level figure. Are your categorical variables consistent in their spelling? Did you check that your European collaborator used a dot instead of a comma to delineate decimal points? Will you extract every statistic and string them together as your axis title for a 6-panel figure *by hand*? This is where understanding how to manipulate strings *en masse* is helpful. 

String manipulation is a fundamental skill in every programming language, and every language has its perks and quirks in handling strings. The common thread (pun intended) that ties them together is regular expressions which form the basis for patterns and their related operations. In bash, you may have used them to find or substitute parts of a string using `grep` or `sed`/`awk`. Here, we will cover some regular expressions as part of the examples and illustrate how they are interpreted in R. However, the topic of regular expression itself is vast. Please refer to the second page of the [stringr cheat sheet](https://github.com/rstudio/cheatsheets/blob/main/strings.pdf) for comprehensive coverage of most of the regex in R.

!!! note "Biological strings"

    At the fundamental level, all biological sequences are strings. They can be handled with the functions covered here, but more is needed to extract biological relevance from them. When we think about biological sequences such as nucleic or amino acid sequences, we are interested in the relationships within and between strings (translations, alignments, *k*-mers, secondary structures, etc.). R has some packages that can handle such operations (e.g., seqinr, Biostrings, DADA2, see other packages in the Bioconductor repository). However, if these analyses are your primary interest, I suggest you look into other programming languages due to the inefficiencies of R in this regard.

## A brief detour

This lesson will take a brief sojourn from *exploratory data analyses* and explore the realm of string-based methods in R via the **stringr** package. For this lesson, we will constrain our explorations using the following biological context: 

> A characteristic of estuarine sediment (where this data originated from) is its active nitrogen-cycling members of the prokaryotic community. For those participating in ammonia or nitrite oxidation, their lineages often have the prefix "Nitroso-" for ammonia oxidisers and "Nitro-" for nitrite oxidisers. However, specific lineages within these groups (assigned at varying taxonomic levels) bear the prefix but do not harbour the necessary genetic machinery for this function.

The above forms the boundary around the data which we will use for this lesson. Here, we will explore a group of functions at an abstract level. We aim to provide awareness to a set of tools to manipulate (i.e., search, subset, substitute) strings. Alongside that, we want you to be comfortable in constructing regular expressions. This can be daunting if you are new to them. However, we like you to keep in mind that it is both an engineering problem (maximising efficiency with pattern captures) and an art form (being creative and playful to adapt to situations). We stress that there are multiple ways (even within the same framework) to obtain similar results. Some functions are Swiss-army knives when paired with the right regex patterns.

## Subset a character vector

Lets prepare a character vector for us to work on.

!!! r-project "code"

    ```r
    nitros <- str_subset(tax$Taxon, "Nitro")
    head(nitros)
    ```

    > ```
    > [1] "d__Bacteria; p__Nitrospirota; c__Nitrospiria; o__Nitrospirales; f__Nitrospiraceae; g__Nitrospira; s__uncultured_Cytophaga"            
    > [2] "d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Nitrosococcales; f__Nitrosococcaceae; g__SZB85; s__uncultured_bacterium"   
    > [3] "d__Archaea; p__Crenarchaeota; c__Nitrososphaeria; o__Nitrosopumilales; f__Nitrosopumilaceae; g__Candidatus_Nitrosopumilus"            
    > [4] "d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Nitrosococcales; f__Nitrosococcaceae; g__SZB85; s__uncultured_bacterium"   
    > [5] "d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Nitrosococcales; f__Nitrosococcaceae; g__SZB85; s__uncultured_bacterium"   
    > [6] "d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Nitrosococcales; f__Nitrosococcaceae; g__FS142-36B-02; s__uncultured_gamma"
    > ```

The above sets us off with a vector subset that returns a character object `nitros` for every element where the pattern `"Nitro"` occurs in. This uses a sub-string as a pattern.

??? tip "Equivalent code"

    === "Vector subset (tidyverse)"

        ```r
        tax$Taxon[str_detect(tax$Taxon, "Nitro")]
        ```

    === "Pattern matching (base R)"

        ```r
        grep(pattern = "Nitro", x = tax$Taxon, value = TRUE)
        ```

!!! note "Anatomy of pattern-based string functions in **stringr**"

    All stringr functions start with the string to operate on as the first argument. For functions that work with regex, the second argument is always the pattern. For functions that perform replacements, the pattern used for replacement is the third argument. For functions that perform searches, the third argument is usually `negate = <Boolean>` which when `TRUE` does an inverse search of the pattern along the string.

    ```
    str_*(<string>, <pattern>, <replacement>|<negate>)
    ```

## Extract sub-string

What lineages do our ammonia and nitrite oxidisers come from? We can inspect their assignment at various taxonomic levels to find out.

!!! r-project "code"    

    ```r
    # Extract domain-level assignments
    str_extract(nitros, "d__[^;]+") %>%
      unique()

    # Extract phylum-level assignments
    str_replace(nitros, ".+p__([^;]+);.+", "Phylum: \\1") %>%
      unique()
    ```

    > ```
    > [1] "d__Bacteria" "d__Archaea"
    > 
    > [1] "Phylum: Nitrospirota"   "Phylum: Proteobacteria" "Phylum: Crenarchaeota"  "Phylum: Nitrospinota"  
    > [5] "Phylum: Zixibacteria"  
    > ```

Here we used two different ways of getting sub-strings out of long strings. 

1. `str_extract()` finds `d__` followed by anything that is NOT as ";" (`[^;]`) more than once (`+`). Note that the taxonomic levels are delineated using a semicolon. When it finds a match to a semicolon, it will stop and return the resulting sub-string. The text below illustrates the pattern matching where `+` represents "still matching" and `!` represents "not a match".
```
string:     d__Bacteria; p__Nitrospirota; c__Nitrospiria; o__Nitrospirales; f__Nitrospiraceae; g__Nitrospira; s__uncultured_Cytophaga
extract:    d__[^;]++++!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
2. `str_replace` looks at everything `.+` before `p__` then captures the pattern "anything except a semicolon more than once" `[^;]+`. This captured pattern ends with a semicolon and everything else in the string *et cetera*. We replace the pattern with `Phylum: ` then the captured pattern represented in regex by `\\1` (meaning captured pattern 1). The text below illustrates the matched string, where "everything else" is represented using `*`.
```
string:     d__Bacteria; p__Nitrospirota; c__Nitrospiria; o__Nitrospirales; f__Nitrospiraceae; g__Nitrospira; s__uncultured_Cytophaga
match:      *************p__[^;]++++++++;********************************************************************************************
capture:    !!!!!!!!!!!!!!!!Nitrospirota!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

Lets attempt to capture more than one pattern:

!!! r-project "code"

    ```r
    # Extract phylum and genus level assignments and sort unique outputs
    str_replace(nitros, ".+p__([^;]+);.+g__([^;]+).+", "Phylum: \\1, Genus: \\2") %>% 
      unique() %>% 
      str_sort()
    ```

    > ```
    >  [1] "d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Nitrococcales; f__Nitrococcaceae"    
    >  [2] "d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Nitrosococcales; f__Nitrosococcaceae"
    >  [3] "Phylum: Crenarchaeota, Genus: Candidatus_Nitrosopelagicus"                                      
    >  [4] "Phylum: Crenarchaeota, Genus: Candidatus_Nitrosopumilus"                                        
    >  [5] "Phylum: Crenarchaeota, Genus: Nitrosopumilaceae"                                                
    >  [6] "Phylum: Nitrospinota, Genus: P9X2b3D02"                                                         
    >  [7] "Phylum: Nitrospirota, Genus: 4-29-1"                                                            
    >  [8] "Phylum: Nitrospirota, Genus: Nitrospira"                                                        
    >  [9] "Phylum: Nitrospirota, Genus: uncultured"                                                        
    > [10] "Phylum: Proteobacteria, Genus: 3PJM14"                                                          
    > [11] "Phylum: Proteobacteria, Genus: AqS1"                                                            
    > [12] "Phylum: Proteobacteria, Genus: Cm1-21"                                                          
    > [13] "Phylum: Proteobacteria, Genus: FS142-36B-02"                                                    
    > [14] "Phylum: Proteobacteria, Genus: IS-44"                                                           
    > [15] "Phylum: Proteobacteria, Genus: MSB-1D1"                                                         
    > [16] "Phylum: Proteobacteria, Genus: Nitrosomonas"                                                    
    > [17] "Phylum: Proteobacteria, Genus: SZB85"                                                           
    > [18] "Phylum: Zixibacteria, Genus: Zixibacteria"
    > ```

Let's break a few items down:

1. We are now capturing two patterns: `\\1` represents the phylum name, and `\\2` represents the genus name. The numbers represent the order of the captured pattern that occurs in the string.
2. Some of it did not work, resulting in the entire string being printed out (the default behaviour in many of the stringr functions, better safe than sorry!)

??? question "How would you fix the issue in point 2?"

    Add a `str_subset()` prior to `str_replace()` and use the genus prefix `g__` to retain all taxonomies with valid genera assignments.

Notice the subtle change in regex of `.+` to `.*`? Where `.+` means "anything more than once", `.*` means "anything more than zero times". Although the change is subtle enough that it will provide the desired output most times. If you did not use `*`, the output will look slightly weird where there are seemingly duplicates but one of them missing a letter at the end. Try it to see for your self!  

## Detecting and counting patterns

How many of these lineages are putative ammonia oxidisers (i.e., "Nitroso-")? Lets count them:

!!! r-project "code"

    ```r
    str_detect(nitros, "Nitroso") %>%
      sum()
    ```

    > ```
    > [1] 75
    > ```

If we only ran the `str_detect()` part, it will return a Boolean logical vector. The pipe to sum counts the number of `TRUE` elements. 

## Concatenate strings



<!--

In the previous lesson, we created a table of elements in `tax` but not in `asv`. This information resides in the `tax_asv_anti$Taxon`, where we can find the taxonomic lineage of sequence variants that were not part of the count table. If we are to find out what was left out of the count table, we will need to inspect this column.

!!! r-project "code"

    ```r
    head(tax_asv_anti$Taxon)
    ```

    > ```
    > [1] "d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast"                                   
    > [2] "d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast"                                   
    > [3] "d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast"                                   
    > [4] "d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast"                                   
    > [5] "d__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__Mitochondria; g__Mitochondria; s__uncultured_bacterium"
    > [6] "d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast"
    > ```

The first few lines are plastids (i.e., chloroplasts and mitochondria). These are common and often unwanted sequences found in 16S rRNA amplicon studies. Many primers that flank and amplify the 16S rRNA hypervariable regions also amplify homologous sequence regions in chloroplast 16S and mitochondrial 18S rRNA sequences.

!!! r-project "code"

    ```r
    tail(tax_asv_anti$Taxon)
    ```

    > ```
    > [1] "d__Bacteria; p__Bacteroidota; c__Bacteroidia; o__Bacteroidales; f__Marinilabiliaceae; g__Carboxylicivirga"                   
    > [2] "d__Bacteria; p__Acidobacteriota; c__Thermoanaerobaculia; o__Thermoanaerobaculales; f__Thermoanaerobaculaceae; g__Subgroup_10"
    > [3] "d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast"                            
    > [4] "d__Bacteria; p__Acidobacteriota; c__Thermoanaerobaculia; o__Thermoanaerobaculales; f__Thermoanaerobaculaceae; g__Subgroup_10"
    > [5] "d__Bacteria; p__Cyanobacteria; c__Cyanobacteriia; o__Chloroplast; f__Chloroplast; g__Chloroplast"                            
    > [6] "d__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Steroidobacterales; f__Woeseiaceae; g__Woeseia" 
    > ```

There are also sequences assigned to non-organellar prokaryotic lineages. 

Given our current information, how many excluded sequence variants were plastids? To answer that, we will need to:

* Detect the words "Chloroplast" and "Mitochondria" occur in the column
* Count the frequency

!!! r-project "code"

    ```r
    # Detect chloroplast
    chloroplast <- str_detect(tax_asv_anti$Taxon, "Chloroplast")
    # Detect mitochondria
    mitochondria <- str_detect(tax_asv_anti$Taxon, "Mitochondria")

    # Count number of TRUE statements
    sum(chloroplast)
    sum(mitochondria)
    ```

    > ```
    > [1] 115
    > [1] 36
    > ```

In the block above, we used `str_detect()`. It searches a character vector (first argument: `string`) for matches to a regex pattern (second argument: `pattern`) and returns a Boolean logical vector. It can also do an inverse search if the third argument (`negate = TRUE`) is provided. Given that the returned vector is Boolean, we feed it to `sum()` to count the number of `TRUE` in the vector.

We can also combine the regex pattern to search for both `"Chloroplast"` and `"Mitochondria"`:

!!! r-project "code"

    ```r
    str_detect(tax_asv_anti$Taxon, "Chloroplast|Mitochondria") %>%
      sum()
    ```

    > ```
    > [1] 151
    > ```

The vertical bar `|` functions as an "OR" statement, meaning our pattern represents: `"Chloroplast"` OR `"Mitochondria"`.

!!! warning "`str_count()` vs `str_detect() %>% sum()`"

    There is another function `str_count()` in the stringr package. The description on the help page for this function states: "Counts the number of times pattern is found within each element of string." This is not the same as counting the frequency of occurrence in a vector using `str_detect() %>% sum()`. Above, you have seen what `str_detect() %>% sum()` is doing. Below is an example using `"Chloroplast"` as the pattern to count:

    !!! r-project "code"

        ```r
        str_count(tax_asv_anti$Taxon, "Chloroplast")
        ```

        > ```
        >   [1] 3 3 3 3 0 3 0 3 3 3 0 3 0 3 0 3 3 3 0 3 3 0 0 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 0 0 3 3 3
        >  [57] 0 3 3 3 3 3 3 3 0 3 3 0 0 3 3 0 0 3 3 3 3 3 3 0 3 0 3 0 0 3 0 0 0 3 3 3 3 3 3 3 3 0 0 3 0 3 0 3 3 0 0 3 0 0 3 3
        > [113] 0 3 3 3 3 3 3 0 3 3 3 3 3 3 3 3 3 3 3 3 3 0 0 0 0 0 0 0 0 0 0 0 0 3 3 3 3 3 3 3 3 3 3 3 0 3 3 0 0 0 0 0 0 0 0 0
        > [169] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 0 3 0
        > ```
    
    Notice that it is counting the number of occurrence of the pattern `"Chloroplast"` *IN EACH ELEMENT* (i.e., `"Chloroplast"` occurs three times in each string at the order, family, and genus level)

## Subset and replace strings



-->


!!! tip "Check your patterns"

    How do you know whether your regex-based string operation(s) ran successfully? 


??? tip "Extra: Text-to-column"

    We can also separate the `Taxon` column into individual columns to a get a better sense of its contents via `View()`.

    !!! r-project "code"

        ```r
        taxa_levels <- c("domain", "phylum", "class", "order", "family", "genus", "species")
        tax_asv_anti_sep <- separate(
          tax_asv_anti, 
          col = "Taxon", 
          into = taxa_levels, 
          sep = "; ", 
          fill = "right"
        )
        ```

        > ```
        >                         Feature_ID      domain            phylum                  class            order
        > 1 72211f0b523273328bffca2f097dfacb d__Bacteria  p__Cyanobacteria      c__Cyanobacteriia   o__Chloroplast
        > 2 71e754dc2ffb103fc01f8badad649f08 d__Bacteria  p__Cyanobacteria      c__Cyanobacteriia   o__Chloroplast
        > 3 e8b50cb415b8aa00e1b00f12a17180ed d__Bacteria  p__Cyanobacteria      c__Cyanobacteriia   o__Chloroplast
        > 4 c45d2e13acee9d895e2000418193abbf d__Bacteria  p__Cyanobacteria      c__Cyanobacteriia   o__Chloroplast
        > 6 1d729775ed6c5422f9c2e8033e06d127 d__Bacteria p__Proteobacteria c__Alphaproteobacteria o__Rickettsiales
        > 7 c71c662746183cec5fd2023b1b1c56b0 d__Bacteria  p__Cyanobacteria      c__Cyanobacteriia   o__Chloroplast
        >            family           genus                 species Confidence
        > 1  f__Chloroplast  g__Chloroplast                    <NA>  1.0000000
        > 2  f__Chloroplast  g__Chloroplast                    <NA>  1.0000000
        > 3  f__Chloroplast  g__Chloroplast                    <NA>  1.0000000
        > 4  f__Chloroplast  g__Chloroplast                    <NA>  0.9999952
        > 6 f__Mitochondria g__Mitochondria s__uncultured_bacterium  0.9532516
        > 7  f__Chloroplast  g__Chloroplast                    <NA>  1.0000000
        > ```
    
    The arguments are as follows:

    1. Data frame to split (`tax_asv_anti`)
    2. Column to split ("Taxon")
    3. Columns to split `Taxon` into (based on `taxa_levels`)
    4. Delimiter to split based on
    5. If there are insufficient elements, add `NA` to the:
           * `"right"`
           * `"left"`
           * right, and `"warn"` about insufficient elements (default behaviour)
    
    The inverse operation is called `unite()`.

